src/transactions/TransactionUtils.cpp

```diff
@@ -1172,6 +1174,321 @@ removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx, AccountID const& account,
+            case SponsorshipResult::TOO_MANY_SPONSORED:
+                // This is impossible because there's no sponsored account.
+                // Fall through and throw.
+            case SponsorshipResult::TOO_MANY_SUBENTRIES:
+                // This is impossible because claimable balances don't use
+                // subentries. Fall through and throw.
+            default:
+                throw std::runtime_error("Unexpected result from "
+                                         "canEstablishEntrySponsorship");
+            }
+
+            ltxInner.create(claimableBalance);
+
+            return true;
+        };
+
+        // use a lambda so we don't hold a reference to the
+        // LiquidityPoolEntry
+        auto constantProduct = [&]() -> auto&
+        {
+            return pool.current().data.liquidityPool().body.constantProduct();
+        };
+
+        if (balance != 0)
+        {
+            auto amountA = getPoolWithdrawalAmount(
+                balance, constantProduct().totalPoolShares,
+                constantProduct().reserveA);
+
+            auto amountB = getPoolWithdrawalAmount(
+                balance, constantProduct().totalPoolShares,
+                constantProduct().reserveB);
+
+            if (((amountA != 0 &&
+                  !redeemIntoClaimableBalance(constantProduct().params.assetA,
+                                              amountA)) ||
+                 (amountB != 0 &&
+                  !redeemIntoClaimableBalance(constantProduct().params.assetB,
+                                              amountB))))
+            {
+                return false;
+            }
+
+            constantProduct().totalPoolShares -= balance;
+            constantProduct().reserveA -= amountA;
+            constantProduct().reserveB -= amountB;
+        }
+
+        decrementLiquidityPoolUseCount(
+            ltxInner, constantProduct().params.assetA, accountID);
+        decrementLiquidityPoolUseCount(
+            ltxInner, constantProduct().params.assetB, accountID);
+        decrementPoolSharesTrustLineCount(pool);
+    }
+    ltxInner.commit();
+
+    return true;
+}
+
+void
+decrementPoolSharesTrustLineCount(LedgerTxnEntry& liquidityPool)
+{
+    auto poolTLCount = --liquidityPool.current()
+                             .data.liquidityPool()
+                             .body.constantProduct()
+                             .poolSharesTrustLineCount;
+    if (poolTLCount == 0)
+    {
+        liquidityPool.erase();
+    }
+    else if (poolTLCount < 0)
+    {
+        throw std::runtime_error("poolSharesTrustLineCount is negative");
+    }
+}
+
+void
+decrementLiquidityPoolUseCount(AbstractLedgerTxn& ltx, Asset const& asset,
+                               AccountID const& accountID)
+{
+    if (!isIssuer(accountID, asset) && asset.type() != ASSET_TYPE_NATIVE)
+    {
+        auto assetTrustLine = ltx.load(trustlineKey(accountID, asset));
```

Comment:
For the usage of this method during auth revocation, it would deactivate the entries loaded with `loadPoolShareTrustLinesByAccountAndAsset`. This is a good point though.